File: pdi/docs/First_steps/hello_data.f90
!*******************************************************************************
! Copyright (C) 2026 Julien Bigot <julien@julien-bigot.fr>
! All rights reserved.
!
! Redistribution and use in source and binary forms, with or without
! modification, are permitted provided that the following conditions are met:
! * Redistributions of source code must retain the above copyright notice,
!   this list of conditions and the following disclaimer.
! * Redistributions in binary form must reproduce the above copyright notice,
!   this list of conditions and the following disclaimer in the documentation
!   and/or other materials provided with the distribution.
! * Neither the names of CEA, nor the names of the contributors may be used to
!   endorse or promote products derived from this software without specific
!   prior written  permission.
!
! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
! AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
! ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
! LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
! CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
! SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
! INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
! CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
! ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
! POSSIBILITY OF SUCH DAMAGE.
!******************************************************************************/

!! [example]
program hello_data
    use paraconf
    use PDI
    implicit none

    type(PC_tree_t),target :: conf
    integer                :: my_world

    call PC_parse_path("hello_data.yml", conf)
    call PDI_init(conf)

    my_world = 42
    call PDI_share("world", my_world, PDI_OUT)
    !variable my_world is shared with PDI

    call PDI_reclaim("world");
    !variable my_world is no longer shared with PDI

    call PDI_finalize()

end program hello_data
!! [example]

File: pdi/docs/First_steps/hello_data.c
/*
 * SPDX-FileCopyrightText: 2021-2026 Commissariat a l'energie atomique et aux energies alternatives (CEA)
 * SPDX-FileCopyrightText: 2024 Commissariat a l'energie atomique et aux energies alternatives (CEA)
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include <pdi.h>

int main(int argc, char* argv[])
{
	PDI_init(PC_parse_path("hello_data.yml"));
	int my_world = 42;
	PDI_share("world", &my_world, PDI_OUT);
	//variable my_world is shared with PDI

	PDI_reclaim("world");
	//variable my_world is no longer shared with PDI

	PDI_finalize();
	return 0;
}
